src/docs/tree_widget.md

```diff
@@ -0,0 +1,723 @@
+
+```typescript
+        ...
+        EXAMPLE_DATA.map(item => this.itemFactory.toTreeNode(item))
+            .forEach(node => CompositeTreeNode.addChild(root, node));
+        this.tree.root = root;
+````
+
+and implement the dynamic resolution of child nodes by subclassing `TreeImpl`:
+
+```typescript
+export class TreeviewExampleTree extends TreeImpl {
+    @inject(TreeViewExampleTreeItemFactory) private readonly itemFactory: TreeViewExampleTreeItemFactory;
+
+    override async resolveChildren(parent: CompositeTreeNode): Promise<TreeNode[]> {
+        // root children are initialized once and never change
+        if (parent.id === ROOT_NODE_ID) {
+            return [...parent.children];
+        }
+
+        // non-container classes do not have children
+        if (!ExampleTreeNode.is(parent)) {
+            return [];
+        }
+
+        // (optional) caching: if the children are resolved and no child was added/removed, reuse 
+        if (parent.children.length === parent.data.children?.length) {
+            return [...parent.children];
+        }
+
+        // simulate asynchronous loading of children
+        await wait(2000);
+        return (parent.data.children ?? []).map(i => this.itemFactory.toTreeNode(i));        
+    }
+}
+```
+
+Under the hood, the tree expansion is handled by the `TreeExpansionServiceImpl` implementation which maintains the `expanded` state of `ExpandableTreeNode`, sends events related to expanding and collapsing subtrees, and also calls `Tree.refresh()` for the expanded node. The `refresh()` method in turn calls `resolveChildren()` which can be implemented to asyncronously provide the child nodes for the given parent. As a bonus, `Tree.refresh()` marks the node as `busy` until the promise is resolved, leading to a nice busy marker in the form of a spinning circle if the promise is not resolved within a certain amount of time (800ms). This can be observed in this demo code due to the `wait(2000)` call.
+
+## Tree Items with Checkboxes
+
+The `TreeWidget` framework also supports presenting and maintaining checkboxes for tree items. Adding checkboxes is as simple as initializing the `checkboxInfo` property for the tree nodes in the `TreeViewExampleTreeItemFactory` class:
+
+```typescript
+    return <ExampleTreeLeaf>{
+        id: this.toTreeNodeId(item),
+        ...
+        checkboxInfo: { checked: item.backOrdered }
+    };
+```
+
+To react to the user checking/unchecking a checkbox, for example, to update the underlying model, we can override the `TreeModel.markAsChecked()` method and provide our own implementation before calling `super`:
+
+```typescript
+    override markAsChecked(node: TreeNode, checked: boolean): void {
+        if (ExampleTreeLeaf.is(node)) {
+            node.data.backOrdered = checked;
+        }
+        super.markAsChecked(node, checked);
+    }
+```
+
+> **Note:** At the moment (Theia 1.60.x), there is an issue with the UI in which the
```

Comment:
I'd remove the Theia 1.60.x version here because it might have already been present in previous versions, just referencing the issue below should be enough.